src/llama-graph.cpp

```diff
@@ -920,6 +977,27 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
         selection_probs = logits;
     }
 
+    // select top n_group_exp expert groups
+    if (arch == LLM_ARCH_BAILINGMOE2) {
+        const int64_t n_exp_per_group = n_expert / hparams.n_expert_groups;
+
+        // organize experts into n_expert_groups
+        ggml_tensor * group_scores = ggml_transpose(ctx0, ggml_view_2d(ctx0, selection_probs, hparams.n_expert_groups, n_tokens * n_exp_per_group, selection_probs->nb[1] * n_exp_per_group, 0)); // [n_tokens, n_expert_groups]
+        group_scores = ggml_top_k(ctx0, ggml_cont(ctx0, group_scores), 2); // [2, n_expert_groups]
+        group_scores = ggml_get_rows(ctx0, ggml_reshape_3d(ctx0, ggml_cast(ctx0, group_scores, GGML_TYPE_F32), 1, 2, hparams.n_expert_groups), group_scores); // [1, 2, n_expert_groups]
+        group_scores = ggml_reshape_2d(ctx0, group_scores, 2, hparams.n_expert_groups); // [2, n_expert_groups]
+
+        // get top n_group_exp expert groups
+        group_scores = ggml_transpose(ctx0, ggml_sum_rows(ctx0, group_scores)); // [n_expert_groups, 1]
+        ggml_tensor * expert_groups = ggml_top_k(ctx0, ggml_cont(ctx0, group_scores), hparams.n_group_exp); // [n_group_exp, 1]
+        cb(expert_groups->src[0], "ffn_moe_group_argsort", il);
+        cb(expert_groups, "ffn_moe_group_topk", il);
+
+        // mask out the other groups
+        selection_probs = ggml_map_custom2(ctx0, selection_probs, expert_groups, mask_expert_groups, GGML_N_TASKS_MAX, (void *)(intptr_t)&hparams.n_expert_groups); // [n_expert, n_tokens]
+        cb(selection_probs, "ffn_moe_probs_masked", il);
+    }
```

Comment:
We should avoid this custom op. It's good for prototyping, but it makes the implementation useless since the graph will be split in a lot of splits.

I think you can achieve this with a combination of:
```c
new_probs = ggml_scale(selection_probs, -inf);
[ some combination of repeat of expert_groups or views of the probs if needed ]
ggml_set_rows(new_probs, selection_probs, expert_groups)
```

The idea is to create a new probs vector, initialize it with -inf and then fill the selected probs in it with set_rows.

Reply:
FYI: Here is how I implemented this with `set_rows`. 

https://github.com/foldl/chatllm.cpp/blob/f1cf81398284e39402c06a8f6fc7c7681c6d0ff0/models/bailing.cpp#L129-L155

But, there is a missing block to bridge `top_k` and `set_rows`. This is why I created issue #16001.

Reply:
Yeah, this asymmetry in the data type of `get_rows` and `set_rows` is not great. We should resolve this.